aspnetcore/blazor/forms-validation.md

```diff
@@ -223,12 +230,383 @@ In the following example:
+* The placeholder `{SOLUTION NAMESPACE}` is the solution's namespace.
+
+Place the `Starship` model into the solution's `Shared` project so that both the client and server apps can use the model. Since the model requires data annotations, add a package reference for [`System.ComponentModel.Annotations`](https://www.nuget.org/packages/System.ComponentModel.Annotations) to the `Shared` project's project file:
+
+```xml
+<ItemGroup>
+  <PackageReference Include="System.ComponentModel.Annotations" Version="{VERSION}" />
+</ItemGroup>
+```
+
+Replace the `{VERSION}` placeholder with the correct version of the package.
+
+In the server API project, add a controller to process starship validation requests (`Controllers/StarshipValidation.cs`) and return failed validation messages:
+
+```csharp
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using {SOLUTION NAMESPACE}.Shared;
+
+namespace {SOLUTION NAMESPACE}.Server.Controllers
+{
+    [Authorize]
+    [ApiController]
+    [Route("[controller]")]
+    public class StarshipValidationController : ControllerBase
+    {
+        private readonly ILogger<StarshipValidationController> logger;
+
+        public StarshipValidationController(
+            ILogger<StarshipValidationController> logger)
+        {
+            this.logger = logger;
+        }
+
+        [HttpPost]
+        public Dictionary<string, List<string>> Post(Starship starship)
+        {
+            var dict = new Dictionary<string, List<string>>();
+
+            try
+            {
+                if (starship.Classification == "Defense" && 
+                    string.IsNullOrEmpty(starship.Description))
+                {
+                    dict.Add(nameof(starship.Description),
+                        new List<string>() { "For a 'Defense' ship classification, " +
+                        "'Description' is required." });
+                }
+            }
+            catch (Exception ex)
+            {
+                logger.LogError("Validation Error: {MESSAGE}", ex.Message);
+            }
+
+            return dict;
+        }
+    }
+}
+```
+
+> [!NOTE]
+> The server API returns a *Bad Request - 400* status code when the model fails validation on the server. This behavior is normal for the example shown here because the model that's shared in the hosted Blazor solution has data annotations applied in the `Shared` project.
+
+In the client app, create a server validation component to display error messages returned by the server API (`ServerValidator.cs`):
+
+```csharp
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Forms;
 
-    private async Task<bool> ServerValidate(EditContext editContext)
+namespace {SOLUTION NAMESPACE}.Client
+{
+    public class ServerValidator : ComponentBase
```

Comment:
This is a copy/paste of the BusinessValidator, maybe "refactor" this : one first part on how to show arbitrary error messages, then how to provide client-side error, then how to provide server side  errors.

Reply:
Good idea ... now you see what I meant by "very rough" draft. :smile:

Reply:
No worries, it's also nice to exchange early on these thing.